### PR TITLE
optimize: pooled HeaderScanner and HeaderValueScanner

### LINKED ISSUE #227: :zap: optimize: may HeaderScanner and HeaderValueScanner be pooled?
**Is your feature request related to a problem? Please describe.**

is it possible to make  [HeaderScanner](https://github.com/cloudwego/hertz/blob/develop/pkg/protocol/http1/ext/headerscanner.go#L53)  and [HeaderValueScanner](https://github.com/cloudwego/hertz/blob/develop/pkg/protocol/http1/ext/headerscanner.go#L75) pooled ?

### PR DESCRIPTION
**Pull request by Haswf**
#### What type of PR is this?

optimize: pooled HeaderScanner and HeaderValueScanner

#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) Translate the PR title into Chinese.
optimize: 池化 HeaderScanner and HeaderValueScanner

#### (Optional) More detail description for this PR(en: English/zh: Chinese).

en: use sync.Pool to recycle HeaderScanner and HeaderValueScanner. It also contains unit tests to ensure scanner has been properly reset.
zh(optional): 

#### Which issue(s) this PR fixes:

implements #227

### CODE REVIEW

**Comment by Haswf** [comment]
After further investigation, I realised that the buffer that stores original header content (e.g. `s.B`) is stripped after each `Next` call. The buffer is empty when next returns false, which means essentially we are not avoiding memory allocation by recycling these scanner. Idk if the performance optimization is worth the additional complexity introduced by pooling. We need benchmark to measure and justify performance improvement.

File: pkg/protocol/http1/ext/headerscanner.go [hunk 1]
```diff
@@ -203,7 +218,8 @@ func (s *HeaderValueScanner) next() bool {
 }
 
 func HasHeaderValue(s, value []byte) bool {
-	var vs HeaderValueScanner
+	vs := HeaderValueScannerPool.Get().(*HeaderValueScanner)
+	vs.Reset()
```
Comment (Duslia) [inline pkg/protocol/http1/ext/headerscanner.go:222]: 放回池子的时候再 reset 吧，保证池子里的都是干净的

Comment (Haswf) [reply pkg/protocol/http1/ext/headerscanner.go]: 在defer里面reset和put了

**Comment by li-jin-gou** [comment]
CI need to solve

**Comment by Haswf** [comment]
Some unit tests in `context_test.go` failed when executed locally on go1.18.5. It this relevent to this pr?
> ❯ go test ./...
?       github.com/cloudwego/hertz      [no test files]
?       github.com/cloudwego/hertz/examples/html_rendering      [no test files]
?       github.com/cloudwego/hertz/examples/standard    [no test files]
ok      github.com/cloudwego/hertz/internal/bytesconv   (cached)
?       github.com/cloudwego/hertz/internal/bytestr     [no test files]
?       github.com/cloudwego/hertz/internal/nocopy      [no test files]
ok      github.com/cloudwego/hertz/internal/stats       (cached)
2022/10/08 22:01:30.296245 fs.go:858: [Error] HERTZ: Cannot open file="./file", error=open ./file: no such file or directory
--- FAIL: TestContextRenderFileFromFS (0.00s)
    context_test.go:405: assertion failed, unexpected: text/x-go; charset=utf-8, expected: text/plain; charset=utf-8
2022/10/08 22:01:30.296330 fs.go:858: [Error] HERTZ: Cannot open file="./file", error=open ./file: no such file or directory
--- FAIL: TestContextRenderFile (0.00s)
    context_test.go:422: assertion failed, unexpected: text/x-go; charset=utf-8, expected: text/plain; charset=utf-8
FAIL
FAIL    github.com/cloudwego/hertz/pkg/app      0.049s
ok      github.com/cloudwego/hertz/pkg/app/client       (cached)
?       github.com/cloudwego/hertz/pkg/app/client/discovery     [no test files]
ok      github.com/cloudwego/hertz/pkg/app/client/loadbalance   (cached)
?       github.com/cloudwego/hertz/pkg/app/client/retry [no test files]
ok      github.com/cloudwego/hertz/pkg/app/middlewares/client/sd        (cached)
ok      github.com/cloudwego/hertz/pkg/app/middlewares/server/basic_auth        (cached)
ok      github.com/cloudwego/hertz/pkg/app/middlewares/server/recovery  (cached)
ok      github.com/cloudwego/hertz/pkg/app/server       (cached)
ok      github.com/cloudwego/hertz/pkg/app/server/binding       (cached)
?       github.com/cloudwego/hertz/pkg/app/server/registry      [no test files]
ok      github.com/cloudwego/hertz/pkg/app/server/render        (cached)
ok      github.com/cloudwego/hertz/pkg/common/adaptor   (cached)
ok      github.com/cloudwego/hertz/pkg/common/bytebufferpool    (cached)
ok      github.com/cloudwego/hertz/pkg/common/compress  (cached)
ok      github.com/cloudwego/hertz/pkg/common/config    (cached)
ok      github.com/cloudwego/hertz/pkg/common/errors    (cached)
?       github.com/cloudwego/hertz/pkg/common/hlog      [no test files]
ok      github.com/cloudwego/hertz/pkg/common/stackless (cached)
?       github.com/cloudwego/hertz/pkg/common/test/assert       [no test files]
ok      github.com/cloudwego/hertz/pkg/common/test/mock (cached)
ok      github.com/cloudwego/hertz/pkg/common/timer     (cached)
?       github.com/cloudwego/hertz/pkg/common/tracer    [no test files]
ok      github.com/cloudwego/hertz/pkg/common/tracer/stats      (cached)
?       github.com/cloudwego/hertz/pkg/common/tracer/traceinfo  [no test files]
ok      github.com/cloudwego/hertz/pkg/common/ut        (cached)
ok      github.com/cloudwego/hertz/pkg/common/utils     (cached)
ok      github.com/cloudwego/hertz/pkg/network  (cached)
ok      github.com/cloudwego/hertz/pkg/network/dialer   (cached)
ok      github.com/cloudwego/hertz/pkg/network/netpoll  (cached)
ok      github.com/cloudwego/hertz/pkg/network/standard (cached)
ok      github.com/cloudwego/hertz/pkg/protocol (cached)
?       github.com/cloudwego/hertz/pkg/protocol/client  [no test files]
?       github.com/cloudwego/hertz/pkg/protocol/consts  [no test files]
ok      github.com/cloudwego/hertz/pkg/protocol/http1   (cached)
ok      github.com/cloudwego/hertz/pkg/protocol/http1/ext       (cached)
?       github.com/cloudwego/hertz/pkg/protocol/http1/factory   [no test files]
?       github.com/cloudwego/hertz/pkg/protocol/http1/proxy     [no test files]
ok      github.com/cloudwego/hertz/pkg/protocol/http1/req       (cached)
ok      github.com/cloudwego/hertz/pkg/protocol/http1/resp      (cached)
?       github.com/cloudwego/hertz/pkg/protocol/suite   [no test files]
ok      github.com/cloudwego/hertz/pkg/route    (cached)
?       github.com/cloudwego/hertz/pkg/route/consts     [no test files]
?       github.com/cloudwego/hertz/pkg/route/param      [no test files]
FAIL

**Review by Duslia — APPROVED** [review]

**Comment by li-jin-gou** [comment]
Is it possible to add a benchmark to see the effect?

**Comment by Haswf** [comment]
> Good job and Is it possible to add a benchmark to see the effect? @Haswf

Sure, I will run benchmark against current develop branch.

**Comment by welkeyever** [comment]
Update this PR if there is anything new to the bench test 🙏

**Comment by li-jin-gou** [comment]
benchmark

old:
```
goos: darwin
goarch: amd64
pkg: github.com/cloudwego/hertz
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkRequestHeaderRead-4    	135763893	        94.34 ns/op	       0 B/op	       0 allocs/op
BenchmarkResponseHeaderRead-4   	122723790	        93.91 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/cloudwego/hertz	46.838s
```

new:
```
goos: darwin
goarch: amd64
pkg: github.com/cloudwego/hertz
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkRequestHeaderRead-4    	100000000	       104.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkResponseHeaderRead-4   	120312861	       103.7 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/cloudwego/hertz	36.459s
```

It doesn't seem to be optimised, but we'll have to work out why later and not merge it for now.

**Pull request closed without merging**